sdk/core/azure-core-amqp/src/test/java/com/azure/core/amqp/implementation/MessageFluxTest.java

```diff
@@ -0,0 +1,303 @@
+        final TestPublisher<ReactorReceiver> upstream = TestPublisher.create();
+        final MessageFlux messageFlux = new MessageFlux(upstream.flux(), prefetch, creditFlowMode, retryPolicy);
+
+        final ReactorReceiver receiver = mock(ReactorReceiver.class);
+        when(receiver.receive()).thenReturn(Flux.never());
+        when(receiver.getEndpointStates()).thenReturn(Flux.never());
+        when(receiver.closeAsync()).thenReturn(Mono.empty());
+
+        StepVerifier.create(messageFlux)
+            .then(() -> upstream.next(receiver))
+            .then(() -> upstream.error(error))
+            .verifyError();
+
+        verify(receiver).closeAsync();
+        upstream.assertCancelled();
+    }
+
+    @ParameterizedTest
+    @CsvSource({
+        "EmissionDriven,1",
+        "RequestDriven,0"
+    })
+    public void shouldCloseReceiverWhenDownstreamSignalsCancel(CreditFlowMode creditFlowMode, int prefetch) {
+        final TestPublisher<ReactorReceiver> upstream = TestPublisher.create();
+        final MessageFlux messageFlux = new MessageFlux(upstream.flux(), prefetch, creditFlowMode, retryPolicy);
+
+        final ReactorReceiver receiver = mock(ReactorReceiver.class);
+        when(receiver.receive()).thenReturn(Flux.never());
+        when(receiver.getEndpointStates()).thenReturn(Flux.never());
+        when(receiver.closeAsync()).thenReturn(Mono.empty());
+
+        StepVerifier.create(messageFlux)
+            .then(() -> upstream.next(receiver))
+            .thenCancel()
+            .verify();
+
+        verify(receiver).closeAsync();
+        upstream.assertCancelled();
+    }
+
+    @ParameterizedTest
+    @CsvSource({
+        "EmissionDriven,1",
+        "RequestDriven,0"
+    })
+    public void shouldTerminateWhenReceiverEmitsNonRetriableError(CreditFlowMode creditFlowMode, int prefetch) {
+        final AmqpException error = new AmqpException(false, "non-retriable", null);
+        final AmqpRetryPolicy retryPolicy = new FixedAmqpRetryPolicy(new AmqpRetryOptions());
+        final TestPublisher<ReactorReceiver> upstream = TestPublisher.create();
+        final MessageFlux messageFlux = new MessageFlux(upstream.flux(), prefetch, creditFlowMode, retryPolicy);
+
+        final ReactorReceiver receiver = mock(ReactorReceiver.class);
+        when(receiver.receive()).thenReturn(Flux.empty());
+        when(receiver.getEndpointStates()).thenReturn(Flux.error(error));
+        when(receiver.closeAsync()).thenReturn(Mono.empty());
+
+        StepVerifier.create(messageFlux)
+            .then(() -> upstream.next(receiver))
+            .verifyErrorMatches(e -> e == error);
+
+        // Expecting closeAsync invocation from two call sites -
+        // 1. before the retry to obtain the next receiver.
+        // 2. when operator terminates due to non-retriable error.
+        verify(receiver, new AtLeast(2)).closeAsync();
+        upstream.assertCancelled();
+    }
+
+    @ParameterizedTest
+    @CsvSource({
```

Comment:
Similar comment about using a method source instead of this so there is 1 place to change if we end up updating/adding these.

Reply:
**[DONE]**